types/index.d.ts

```diff
@@ -0,0 +1,31 @@
+import { Readable } from "stream";
+
+export type FormMethodOptions = {
+  readonly payload?: string,
+  readonly headers?: string
+}
+
+type FormMethodDefaultOptions = {
+  readonly payload?: 'payload',
+  readonly headers?: 'headers'
+}
+
+type ComputedFormProperty<T extends FormMethodOptions, Property extends keyof T, DefaultValue extends string, ReturnType> = {
+    [K in T[Property] as keyof T[Property] extends never 
+    ? DefaultValue
+    : K extends PropertyKey 
+        ? K 
+        : never
+    ]: ReturnType
+}
+
+type FormMethodResult<T extends FormMethodOptions, K extends keyof T = keyof T> = string extends T[K]
+  ? ComputedFormProperty<T, 'headers', string, Record<string, string> | undefined> | ComputedFormProperty<T, 'payload', string, Readable | undefined>
+  : ComputedFormProperty<T, 'headers', 'headers', Record<string, string>> & ComputedFormProperty<T, 'payload', 'payload', Readable>
+
+/**
+ * @param {Record<string, unknown>}  json -  A JSON object that defines the fields of the form.
+ * @param {FormMethodOptions}  opts - An object containing properties to modify the output field names.
+ * @returns {FormMethodResult} A JSON object with a payload field representing the data stream and a headers field containing the content-type set to "application/json".
+ */
+export default function formMethod<const T extends FormMethodOptions = FormMethodDefaultOptions>(json: Record<string, unknown>, opts?: T): FormMethodResult<T>
```

Comment:
This is working (note the wrong option):

```
  const myForm = formAutoContent({
    field1: 'value1',
    field2: ['value2', 'value2.2']
  }, { headers: 'head', foo: '' });
```

Could we handle it?

Reply:
I found a way to handle this case but with a con: TS won't be able to infer the correct type when using an inline object without a const assertion because the type is widened when using the helper `WithoutExtraProperties`.

However, I realized this could be 'good' as the feature I'm using to infer the types in inline objects is a [new feature from TS 5.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-rc/#const-type-parameters), so without this feature we can support pre TS 5.0 codebases

Reply:
Thanks for looking into it!